kommunicateui/src/main/java/com/applozic/mobicomkit/uiwidgets/conversation/adapter/DetailedConversationAdapter.java

```diff
@@ -1333,9 +1342,54 @@ public boolean onMenuItemClick(MenuItem menuItem) {
                     view.dispatchTouchEvent(MotionEvent.obtain(0, 0, MotionEvent.ACTION_CANCEL, 0f, 0f, 0));
                 }
             });
+
+            if (showSourceURls(message, position)) {
+                String json = message.getMetadata().get(SOURCE_URL);
+                if(json == null) {
+                    return;
+                }
+                myHolder.sourceText.setVisibility(View.VISIBLE);
+                myHolder.urlsLayout.setVisibility(View.VISIBLE);
+                myHolder.sourceUrlDivider.setVisibility(View.VISIBLE);
+
+                Type listType = new TypeToken<List<SourceUrl>>() {}.getType();
+                List<SourceUrl> links = GsonUtils.getObjectFromJson(json, listType);
+
+                myHolder.urlsLayout.removeAllViews();
+                for(SourceUrl url : links) {
+                    SpannableString content = new SpannableString(url.getTitle());
+                    content.setSpan(new UnderlineSpan(), 0, content.length(), 0);
+
+                    View view = LayoutInflater
+                            .from(myHolder.urlsLayout.getContext())
+                            .inflate(R.layout.km_link_view, null, false);
+
+                    TextView tv =  view.findViewById(R.id.src_urls_textview);
+                    tv.setText(content);
+
+                    view.setOnClickListener((data) -> {
+                        openURL(view.getContext(), url.getUrl());
+                    });
+                    myHolder.urlsLayout.addView(view);
+                }
+            }
         }
     }
 
+    private void openURL(Context context, String url) {
+        Uri uri = Uri.parse(url);
+        Intent intent = new Intent(Intent.ACTION_VIEW, uri);
+        context.startActivity(intent);
+    }
+
+    private boolean showSourceURls(Message message, int position) {
+        return message.getContentType() == 0
+                && getItemViewType(position) == 0
+                && !message.getMessage().isEmpty()
+                && message.getType() == 4
+                && message.getMetadata().containsKey(SOURCE_URL);
```

Comment:
Any chance for metadata to be null? Might through NLP

Reply:
Yes, I'll add the check here for meta data null.

Reply:
Done, Added in [ba472f4](https://github.com/Kommunicate-io/Kommunicate-Android-Chat-SDK/pull/535/commits/ba472f4632f8c8cdbf258effab34b649037c322c)